File: client/iosApp/iosApp/screen/chat/ChatViewModel.swift
import Foundation
import chat
import logging

class ChatViewModel : ObservableObject {
    
    var  logger : Logger
    
    var chatRepository : ChatRepository
    
    init(chatRepository: ChatRepository = ChatRepositoryImpl.init(chatApi: ChatApi()) ){
        self.chatRepository = chatRepository
        
        self.logger = Logger.init()
        self.logger.i(message: "ChatViewModel Launched", tag: "ChatViewModel" )
        
        self.initiateConnection()
    }
    
    @Published var messages : [ChatMessage] = []
    
    func initiateConnection(){
        let usernames = ["iPad", "Mac", "iPod", "Mac-Mini", "Watch"]
        self.chatRepository.openConnection(username: usernames.randomElement() ?? "") { exception  in
            if let error = exception?.localizedDescription {
                return
            }
            DispatchQueue.main.async {
                self.observeMessages()
            }
        }
    }
    
    @Published var error : String? = nil
    
    func observeMessages(){
        self.logger.i(message: "Listening for new messages", tag: "ChatViewModel" )
        
        
        self.chatRepository.listenMessagesIos { chatMessage in
            self.messages.append( chatMessage)
        } completionHandler: {  exception  in
            if let error = exception?.localizedDescription {
                self.error = "Failed to listen to messages: " + error
            }
        }
        
    }
    
    @Published var newMessage : String = ""
    
    @Published var sendingMessage : Bool = false
    
    func sendMessage()   -> Void  {
        self.logger.i(message: "Sending message: message -> " + newMessage, tag: "ChatViewModel" )
        
        
        self.error = nil
        self.sendingMessage = true
        self.chatRepository.sendMessage(message: newMessage) { exception  in
            self.sendingMessage = false
            if let error = exception?.localizedDescription {
                self.error =  error
            }
        }
    }
    
    func disconnect()  -> Void  {
        chatRepository.closeConnection {  exception  in
            if let error = exception?.localizedDescription {
                return
            }
        }
    }
}

File: client/androidApp/src/main/java/app/kmmchat/android/screen/chat/ChatViewModel.kt
package app.kmmchat.android.screen.chat

import androidx.compose.runtime.MutableState
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.kmmchat.chat.data.ChatRepositoryImpl
import app.kmmchat.chat.domain.ChatRepository
import app.kmmchat.chat.presentation.ChatScreenEvent
import app.kmmchat.chat.presentation.ChatScreenState
import app.kmmchat.logging.Logger
import kotlinx.coroutines.launch

class ChatViewModel(
    private val chatRepository: ChatRepository = ChatRepositoryImpl()
) : ViewModel() {

    private val _state: MutableState<ChatScreenState> = mutableStateOf(ChatScreenState())
    val state get() = _state


    fun performEvent(event: ChatScreenEvent) {
        val state = state.value
        Logger.v("Performing action: $event on state: $state", "ChatViewModel")

        when (event) {
            is ChatScreenEvent.NavigateUp -> {
                _state.value = state.copy(navigatingUp = event.navigate)
            }
            is ChatScreenEvent.SendMessage -> {
                sendMessage(state.messageText)
            }
            is ChatScreenEvent.DisplayMessage -> {
                _state.value = state.copy(messages = state.messages.plus(event.message))
            }
            is ChatScreenEvent.TypeMessageText -> {
                _state.value = state.copy(messageText = event.text)
            }
            is ChatScreenEvent.OpenConnection -> {
                val usernames = listOf("Alpha", "Beta", "Cupcake", "Donut", "ICS", "Kitkat", "Oreo")
                val username = usernames.random()
                connect(username)
            }
            is ChatScreenEvent.CloseConnection -> {
                disconnect()
            }
            is ChatScreenEvent.ObserveMessages -> {
                listenMessages()
            }
        }
    }


    private fun sendMessage(message: String) = viewModelScope.launch {
        Logger.i("Sending message: content -> $message", "ChatViewModel")

        kotlin.runCatching {
            chatRepository.sendMessage(message)
        }.onFailure {
            Logger.i("Could not send message: stacktrace -> ${it.stackTrace}", "ChatViewModel")
        }
    }


    private fun listenMessages() = viewModelScope.launch {
        Logger.i("Listening for new messages", "ChatViewModel")

        kotlin.runCatching {
            chatRepository.listenMessages().collect {
                Logger.v("New message received: $it", "ChatViewModel")
                performEvent(ChatScreenEvent.DisplayMessage(it))
            }
        }.onFailure {
            Logger.i("Could not observe messages: stacktrace -> ${it.stackTrace}", "ChatViewModel")
        }
    }


    private fun connect(username: String) = viewModelScope.launch {
        Logger.i("Opening chat connection for user: $username", "ChatViewModel")

        kotlin.runCatching {
            chatRepository.openConnection(username)
            performEvent(ChatScreenEvent.ObserveMessages)
        }.onFailure {
            Logger.i("Could not open connection: stacktrace -> ${it.stackTrace}", "ChatViewModel")
        }
    }


    private fun disconnect() = viewModelScope.launch {
        Logger.i("Closing chat connection", "ChatViewModel")

        chatRepository.closeConnection()
    }


    init {
        performEvent(ChatScreenEvent.OpenConnection)
    }

    override fun onCleared() {
        super.onCleared()
        performEvent(ChatScreenEvent.CloseConnection)
    }
}
